tvm_vm/src/executor/zk.rs

```diff
@@ -0,0 +1,822 @@
+
+    let proof = ProofWrapper::deserialize(&proof_as_bytes)?;
+    let public_inputs = FieldElementWrapper::deserialize_vector(&public_inputs_as_bytes)?;
+    let x: Vec<Fr> = public_inputs.iter().map(|x| x.0).collect();
+
+    let vk = if vk_index == 0 {
+        insecure_pvk()
+    } else if vk_index == 1 {
+        global_pvk()
+    } else {
+        my_test_pvk_1()
+    };
+
+    // todo: add alternative for elliptic curve (BLS), read from stack curve id
+
+    let res = Groth16::<Bn254>::verify_with_processed_vk(&vk, &x, &proof.0)
+        .map_err(|e| ZkCryptoError::GeneralError(e.to_string()));
+
+    let succes = res.is_ok();
+    
+    let res = if succes { boolean!(res.unwrap()) } else { boolean!(false) };
+
+    engine.cc.stack.push(res);
+
+    Ok(())
+}
+
+fn pop(barry: &[u8]) -> &[u8; 8] {
+    barry.try_into().expect("slice with incorrect length")
+}
+
+pub(crate) fn execute_poseidon_zk_login(engine: &mut Engine) -> Status {
+    engine.load_instruction(crate::executor::types::Instruction::new("POSEIDON"))?;
+    fetch_stack(engine, 5)?;
+
+    let zkaddr_slice = SliceData::load_cell_ref(engine.cmd.var(0).as_cell()?)?;
+    let zkaddr = unpack_string_from_cell(zkaddr_slice, engine)?;
+
+    let header_and_iss_base64_slice =
+        SliceData::load_cell_ref(engine.cmd.var(1 /* 2 */).as_cell()?)?;
+    let header_and_iss_base64 = unpack_string_from_cell(header_and_iss_base64_slice, engine)?;
+
+    let modulus_slice = SliceData::load_cell_ref(engine.cmd.var(2).as_cell()?)?;
+    let modulus = unpack_data_from_cell(modulus_slice, engine)?;
+
+    let eph_pub_key = engine
+        .cmd
+        .var(3 )
+        .as_integer()?
+        .as_builder::<UnsignedIntegerBigEndianEncoding>(PUBLIC_KEY_BITS)?;
+
+    let eph_pub_key_bytes = eph_pub_key.data();
+
+    let max_epoch_ = engine
+        .cmd
+        .var(4 )
+        .as_integer()?
+        .as_builder::<UnsignedIntegerBigEndianEncoding>( 64)?;
+
+    let max_epoch_bytes = pop(max_epoch_.data());
+    let max_epoch = u64::from_be_bytes(*max_epoch_bytes);
+
+    let public_inputs = calculate_poseidon_hash(
+        &*zkaddr,
+        &*header_and_iss_base64,
+        &eph_pub_key_bytes, 
+        &modulus,
+        max_epoch,
+    )
+    .unwrap();
+
+    let mut public_inputs_as_bytes = vec![];
+    public_inputs.serialize_compressed(&mut public_inputs_as_bytes).unwrap();
+
+    let public_inputs_cell = pack_data_to_cell(&public_inputs_as_bytes, &mut 0).unwrap();
+
+    engine.cc.stack.push(Cell(public_inputs_cell));
+
+    Ok(())
+}
+
+pub fn calculate_poseidon_hash(
+    address_seed: &str,
+    header_and_iss_base64: &str,
+    eph_pk_bytes: &[u8],
+    modulus: &[u8],
+    max_epoch: u64,
+) -> Result<Bn254Fr, ZkCryptoError> {
+    
+    let address_seed = Bn254FrElement::from_str(address_seed).unwrap();
+    let addr_seed = (&address_seed).into();
+
+    let (first, second) = split_to_two_frs(eph_pk_bytes).unwrap();
```

Comment:
Remove pure unwraps, fn should return Err on bad arguments
+ I think it's bad thing to pass json encoded parameters as an argument